### ISSUE TITLE
PV provisioned through CSI plugin does not show up in the pod

### ISSUE THREAD

**Issue by gila** [ubuntu/microk8s#454]
Hi all,

I've been using microk8s for local development and its been really great thus far. One of the things I'm working on is implementing a CSI driver for openEBS that is making use of nvmf. 

I can use microk8s in such a way that the device shows up on the host and then well -- I can do some basic tests with FIO make sure everything is OK and all before I push it up to one of our CI clusters.

Now, with 1.14, I've tried to push the bar a bit and do the whole PV dance through a dynamic provisioner and the CSI driver to see if all works out. I'm close, but at the moment stuck. 

The problem is that everything works, except the volume, is not mounted within the pod (that's an odd statement if you consider the fact that all you want to do is get a volume but bear with me) 

From our CSI node component, we can see that the PVC's are created and destroyed when the pod is destroyed. Here is a snippet of that log file:

```
 NodePublishVolumeRequest { volume_id: "pool/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4", publish_context: {}, staging_target_path: "/var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount", target_path: "/var/snap/microk8s/common/var/lib/kubelet/pods/bdd37985-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount", volume_capability: Some(VolumeCapability { access_mode: Some(AccessMode { mode: SingleNodeWriter }), access_type: Some(Mount(MountVolume { fs_type: "ext4", mount_flags: [] })) }), readonly: false, secrets: {}, volume_context: {"storage.kubernetes.io/csiProvisionerIdentity": "1557001666827-8081-io.openebs.csi-mayastor"} }
[2019-05-04T20:39:48Z TRACE mayastor_grpc::mount] Matched mount: MountInfo { source: "/dev/openebs/pool-pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4", dest: "/var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount", fstype: "ext4", options: ["rw", "relatime"], dump: 0, pass: 0 }
[2019-05-04T20:39:48Z DEBUG mayastor_grpc::mount] Mounting /var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount ...
[2019-05-04T20:39:48Z INFO  mayastor_grpc::mount] Mounted ext4 fs on /var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount with opts "rw,relatime" to /var/snap/microk8s/common/var/lib/kubelet/pods/bdd37985-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount
[2019-05-04T20:39:48Z INFO  mayastor_grpc::node] Published volume pool/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4
```

Two things are most important (for me at least) one is the device `/dev/openebs/PVC-xxx` (PS don't worry no kernel code is abused to create these devices IOW, we entirely run in user space) and the other path is the target path, which is in this case pod "bddd..." and then some. 

This pod should be my pod that runs alpine with fio in it:

```
~kubectl get pod --all-namespaces -o json | jq '.items[] | select(.metadata.uid == "bdd37985-6eac-11e9-a69a-000c290e4fe4")'
{
  "apiVersion": "v1",
  "kind": "Pod",
  "metadata": {
    "creationTimestamp": "2019-05-04T20:39:40Z",
    "name": "fio",
    "namespace": "mayastor",
    "resourceVersion": "68188",
    "selfLink": "/api/v1/namespaces/mayastor/pods/fio",
    "uid": "bdd37985-6eac-11e9-a69a-000c290e4fe4"
  },
  "spec": {
    "containers": [
      {
        "args": [
          "sleep",
          "1000000"
        ],
        "image": "dmonakhov/alpine-fio",
        "imagePullPolicy": "Always",
        "name": "fio",
        "resources": {},
        "terminationMessagePath": "/dev/termination-log",
        "terminationMessagePolicy": "File",
        "volumeMounts": [
          {
            "mountPath": "/volume",
            "name": "ms-volume"
          },
          {
            "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
            "name": "default-token-496kb",
            "readOnly": true
          }
        ]
      }
    ],
```

Ok so that looks to be correct, trust but verify:

```
~pwd
/var/snap/microk8s/common/var/lib/kubelet/pods/bdd37985-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4
~cat vol_data.json | python -m json.tool
{
    "attachmentID": "csi-6cf61499d675cf880cd2bf02fd229288c2b979735ba3321a20218d6e37e98ea5",
    "driverMode": "persistent",
    "driverName": "io.openebs.csi-mayastor",
    "nodeName": "gilatop",
    "specVolID": "pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4",
    "volumeHandle": "pool/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4"
}
```
Ok looks good from here as far as I can tell. I've also matched that with "real" k8s systems in the sense that those where installed by using kubeadm.

Now, onwards to our pod with fio -- there is no mount path with volume there at all, even though I've asked for it:

```
kind: Pod
apiVersion: v1
metadata:
  name: fio
spec:
  volumes:
    - name: ms-volume
      persistentVolumeClaim:
       claimName: ms-volume-claim
  containers:
    - name: fio
      image: dmonakhov/alpine-fio
      args:
        - sleep
        - "1000000"
      volumeMounts:
        - mountPath: "/volume"
          name: ms-volume
```

```
 kubectl get pvc
NAME              STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
ms-volume-claim   Bound    pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4   1Gi        RWO            mayastor       25m
```

To be sure, i've also looked into the mounts of the CSI node, using lsns to figure out the PID and there we find:

```
/dev/openebs/pool-pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4 /var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount ext4 rw,relatime 0 0
/dev/openebs/pool-pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4 /var/snap/microk8s/common/var/lib/kubelet/pods/bdd37985-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount ext4 rw,relatime 0 0
```

When I delete the POD, our CSI implementation -- indeed removes the volume:

```
[2019-05-04T20:39:03Z DEBUG mayastor_grpc] New csi connection
[2019-05-04T20:39:03Z TRACE mayastor_grpc::node] NodeUnpublishVolumeRequest { volume_id: "pool/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4", target_path: "/var/snap/microk8s/common/var/lib/kubelet/pods/63c797da-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount" }
[2019-05-04T20:39:03Z TRACE mayastor_grpc::mount] Matched mount: MountInfo { source: "/dev/openebs/pool-pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4", dest: "/var/snap/microk8s/common/var/lib/kubelet/pods/63c797da-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount", fstype: "ext4", options: ["rw", "relatime"], dump: 0, pass: 0 }
[2019-05-04T20:39:03Z DEBUG mayastor_grpc::node] Unmounting volume pool/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4 at /var/snap/microk8s/common/var/lib/kubelet/pods/63c797da-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount...
[2019-05-04T20:39:03Z DEBUG mayastor_grpc::mount] Unmounting /var/snap/microk8s/common/var/lib/kubelet/pods/63c797da-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount ...
[2019-05-04T20:39:03Z INFO  mayastor_grpc::mount] Filesystem at /var/snap/microk8s/common/var/lib/kubelet/pods/63c797da-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount has been unmounted
[2019-05-04T20:39:03Z INFO  mayastor_grpc::node] Unpublished volume pool/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4 at /var/snap/microk8s/common/var/lib/kubelet/pods/63c797da-6eac-11e9-a69a-000c290e4fe4/volumes/kubernetes.io~csi/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/mount
[2019-05-04T20:39:03Z DEBUG mayastor_grpc] Connection reset by the peer
[2019-05-04T20:39:03Z DEBUG mayastor_grpc] New csi connection
[2019-05-04T20:39:03Z DEBUG mayastor_grpc::node] NodeGetCapabilities request: [GetVolumeStats, StageUnstageVolume]
[2019-05-04T20:39:03Z DEBUG mayastor_grpc] Connection reset by the peer
[2019-05-04T20:39:03Z DEBUG mayastor_grpc] New csi connection
[2019-05-04T20:39:03Z DEBUG mayastor_grpc::node] Unstaging volume pool/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4 at /var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount
[2019-05-04T20:39:03Z TRACE mayastor_grpc::mount] Matched mount: MountInfo { source: "/dev/openebs/pool-pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4", dest: "/var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount", fstype: "ext4", options: ["rw", "relatime"], dump: 0, pass: 0 }
[2019-05-04T20:39:03Z DEBUG mayastor_grpc::mount] Unmounting /var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount ...
[2019-05-04T20:39:03Z INFO  mayastor_grpc::mount] Filesystem at /var/snap/microk8s/common/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-531a5d11-6eac-11e9-a69a-000c290e4fe4/globalmount has been unmounted
```

When looking into the journal however:

```
May 04 21:21:00 gilatop microk8s.daemon-kubelet[7973]: E0504 21:21:00.888610    7973 kubelet_volumes.go:154] Orphaned pod "30537fc0-6e93-11e9-9390-000c290e4fe4" found, but volume paths are still present on disk : There were a total of 1 errors similar to this. Turn up verbosity to see them.
May 04 21:21:02 gilatop microk8s.daemon-kubelet[7973]: E0504 21:21:02.890903    7973 kubelet_volumes.go:154] Orphaned pod "30537fc0-6e93-11e9-9390-000c290e4fe4" found, but volume paths are still present on disk : There were a total of 1 errors similar to this. Turn up verbosity to see them.
May 04 21:21:04 gilatop microk8s.daemon-kubelet[7973]: E0504 21:21:04.889028    7973 kubelet_volumes.go:154] Orphaned pod "30537fc0-6e93-11e9-9390-000c290e4fe4" found, but volume paths are still present on disk : There were a total of 1 errors similar to this. Turn up verbosity to see them.
```

To make this work, to begin with, all I had to do was change some paths in my YAML that pointed to `/var/lib/XXXX` to `/var/snap/microk8s/common/var/lib/XXXX`

So long story short, it seems that somehow, the mount is not propagated within the POD even though everything is there where it should be compared to "normal" k8s. Is there anything I have missed or anything you guys can help with, within the sense of "try this or that"?  

The reason I filed the bug is mostly that I have a *feeling* I'm close to getting it to work, but I do fully understand and appreciate that it's not your goal to run anything and everything from k8s in microk8s.

[inspection-report-20190504_212723.tar.gz](https://github.com/ubuntu/microk8s/files/3144864/inspection-report-20190504_212723.tar.gz)

**Comment by ktsakalozos** [comment]
Hi @gila thank you for opening this issue.

MicroK8s is a certified Kubernetes distribution that ships the actual upstream binaries so it is expected to run anything Kubernetes runs. What we are not covering is all potential configurations the users may expect. We do however provide the means to configure the Kubernetes daemons in any way you please. This is done by editing the files in `/snap/microk8s/current/args/` (there is one file per k8s daemon) and restarting MicroK8s.

I do not have any tip to give you right now, @hyperbolic2346 has worked with CSI and may be able to offer some guidance.

@gila, since I have some space to cover in getting familiar with what you are doing, could you provide me with a minimal setup that would reproduce the error you are seeing? Please be as descriptive as possible. We may be missing some configuration, I see for example this "CSIPersistentVolume" in https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ but I am not sure how to see if it makes any difference.

**Comment by gila** [comment]
@ktsakalozos 

Thank you for the pointer(s) -- I will look into the kubelet and see if I can find something there and will let you know. Thanks!

**Comment by gila** [comment]
@ktsakalozos It seems that the problem was between keyboard and chair. Instead of using `common` I used `current` as the path for creating the gRPC sockets that CSI uses. 

So apologies for the noise, it works end to end now!

**Comment by ktsakalozos** [comment]
I am happy it works now. Should we close this issue?

**Comment by gila** [comment]
Yes sounds good to me :)
